File: Swift/BOJ/BOJ12865.swift
//
//  main.swift
//  BOJalgorithm
//
//  Created by KellyChui on 2023/02/26.
//

import Foundation



let nk = readLine()!.split(separator: " ").map { Int(String($0))! }
var objects: [(weight: Int, value: Int)] = [(0, 0)]
for _ in 0..<nk[0] {
    let wv = readLine()!.split(separator: " ").map { Int(String($0))! }
    objects.append((wv[0], wv[1]))
}

var dpTable = [[Int]](repeating: [Int](repeating: 0, count: nk[1] + 1), count: nk[0] + 1)

for idx in 1...nk[0] {
    for weight in 1...nk[1] {
        if weight - objects[idx].weight >= 0 {
            dpTable[idx][weight] = max(dpTable[idx - 1][weight], dpTable[idx - 1][weight - objects[idx].weight] + objects[idx].value)
        } else {
            dpTable[idx][weight] = dpTable[idx - 1][weight]
        }
    }
}
print(dpTable[nk[0]][nk[1]])

File: Python/BOJ/BOJ12865.py
n, k = map(int, input().split())
wList = []
vList = []
for _ in range(n):
    w, v = map(int, input().split())
    wList.append(w)
    vList.append(v)
dp = [[0] * (k + 1) for _ in range(n + 1)]
for i in range(1, n + 1):
    w, v = wList[i - 1], vList[i - 1]
    for c in range(k + 1):
        dp[i][c] = dp[i - 1][c]
        if c >= w:
            dp[i][c] = max(dp[i][c], dp[i - 1][c - w] + v)
print(dp[n][k])
